pkg/sfu/videolayerselector/dependencydescriptor.go

```diff
@@ -83,129 +83,99 @@ func (d *DependencyDescriptor) Select(extPkt *buffer.ExtPacket, _layer int32) (r
-	if dd.AttachedStructure != nil {
-		// update decode target layer and active decode targets
-		// DD-TODO : these targets info can be shared by all the downtracks, no need calculate in every selector
-		d.updateDependencyStructure(dd.AttachedStructure)
+	if ddwdt.ActiveDecodeTargetsUpdated {
+		d.updateActiveDecodeTargets(*dd.ActiveDecodeTargetsBitmask)
 	}
 
-	// DD-TODO : we don't have a rtp queue to ensure the order of packets now,
-	// so we don't know packet is lost/out of order, that cause us can't detect
-	// frame integrity, entire frame is forwareded, whether frame chain is broken.
-	// So use a simple check here, assume all the reference frame is forwarded and
-	// only check DTI of the active decode target.
-	// it is not effeciency, at last we need check frame chain integrity.
-
-	activeDecodeTargets := dd.ActiveDecodeTargetsBitmask
-	if activeDecodeTargets != nil {
-		d.logger.Debugw("active decode targets", "activeDecodeTargets", *activeDecodeTargets)
+	for _, chain := range d.chains {
+		chain.OnFrame(extFrameNum, fd)
 	}
 
 	// find decode target closest to targetLayer
 	highestDecodeTarget := buffer.DependencyDescriptorDecodeTarget{
 		Target: -1,
 		Layer:  buffer.InvalidLayer,
 	}
-	for _, dt := range ddwdt.DecodeTargets {
-		if dt.Layer.Spatial > d.targetLayer.Spatial || dt.Layer.Temporal > d.targetLayer.Temporal {
-			continue
-		}
-
-		if activeDecodeTargets != nil && ((*activeDecodeTargets)&(1<<dt.Target) == 0) {
-			continue
-		}
-
-		if len(d.structure.DecodeTargetProtectedByChain) == 0 {
-			highestDecodeTarget = dt
-			//d.logger.Debugw("select target", "highestDecodeTarget", highestDecodeTarget, "dtis", fd.DecodeTargetIndications)
-			break
-		}
-
-		if len(d.structure.DecodeTargetProtectedByChain) < dt.Target {
-			// look for lower target
+	var dti dede.DecodeTargetIndication
+	d.decodeTargetsLock.RLock()
+	for _, dt := range d.decodeTargets {
+		if !dt.Active() || dt.Layer.Spatial > d.targetLayer.Spatial || dt.Layer.Temporal > d.targetLayer.Temporal {
 			continue
 		}
 
-		chainIdx := d.structure.DecodeTargetProtectedByChain[dt.Target]
-		if len(fd.ChainDiffs) < chainIdx {
-			// look for lower target
-			continue
+		frameResult, err := dt.OnFrame(extFrameNum, fd)
+		if err != nil {
+			d.decodeTargetsLock.RUnlock()
+			// dtis error, dependency descriptor might lost
+			d.logger.Debugw(fmt.Sprintf("drop packet for frame detection error,  incoming: %v",
+				incomingLayer,
+			), "err", err)
+			d.decisions.AddDropped(extFrameNum)
+			return
 		}
 
-		prevFrameInChain := extFrameNum - uint64(fd.ChainDiffs[chainIdx])
-		if prevFrameInChain != 0 && prevFrameInChain != extFrameNum {
-			if sd, err := d.decisions.GetDecision(prevFrameInChain); err != nil || sd != selectorDecisionForwarded {
-				// look for lower target
-				continue
+		// try to keep lower spatial layer decodable so that we can switch to it seamlessly
+		if frameResult.TargetValid {
+			if highestDecodeTarget.Target == -1 {
```

Comment:
Not sure if this `-1` is special case in DD path. Or maybe, it is representation of `buffer.InvalidLayerSpatial`. If it signifies the invalid layer indicator we use across our SFU code, can you please use that constant?